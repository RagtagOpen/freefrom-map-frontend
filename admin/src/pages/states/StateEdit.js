import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useParams, useLocation } from "react-router-dom";
import _ from 'lodash';

// Redux
import { saveData, getData, selectStates, selectState } from './stateSlice';

function StateEdit() {
    const [saving, setSaving] = useState(false);
    const { id } = useParams();
    const location = useLocation();

    const states = useSelector(selectStates);
    let data = useSelector(selectState(id));

    const [state, setState] = useState({});
    const dispatch = useDispatch();

    useEffect(() => {
        if (states.loaded === false) {
            dispatch(getData());
        }

        if (id === undefined && location.pathname === '/states/new') {
            data = {
                "id": null,
                "name": "",
                "abbrev": "",
                "grade": null
            };
        }

        if (_.isEmpty(state) && !_.isEmpty(data)) {
            setState(_.cloneDeep(data));
        }
    });

    function getValue(e) {
        let value = e.target.value;

        if (e.target.name === "grade") {
            value = parseInt(e.target.value);
        }

        return value;
    }

    const updateState = (e) => {
        const value = getValue(e);

        const updatedState = {
            ...state,
            [e.target.name]: value
        }

        setState(updatedState);
    }

    const saveState = (e) => {
        e.preventDefault();
        e.target.textContent = "Saving..."
        setSaving(true);
        dispatch(saveData(state));
        e.target.textContent = "Saved!"
        setSaving(false);
        e.target.textContent = "Save"
    }

    return (
        <div>
            <nav aria-label="breadcrumb">
                <ol className="breadcrumb">
                    <li className="breadcrumb-item">
                        <a href="/states">State Scores</a>
                    </li>
                    <li className="breadcrumb-item active" aria-current="page">
                        { state.name || "New State" }
                    </li>
                </ol>
            </nav>

            <form>
                <div className="pb-4">
                    <div className="form-group">
                        <label htmlFor={`state-name-${state.id}`}>Name</label>
                        <input
                            id={`state-name-${state.id}`}
                            type="text"
                            className="form-control"
                            name="name"
                            onKeyUp={ updateState }
                            defaultValue={ state.name } />
                    </div>

                    <div className="form-group">
                        <label htmlFor={`state-abbrev-${state.id}`}>Abbreviation</label>
                        <input
                            id={`state-abbrev-${state.id}`}
                            type="text"
                            className="form-control"
                            name="abbrev"
                            onKeyUp={ updateState }
                            defaultValue={ state.abbrev } />
                    </div>

                    <div className="form-group">
                        <label htmlFor={`state-grade-${state.id}`}>Grade</label>
                        <select
                            id={`state-grade-${state.id}`}
                            className="form-control"
                            name="grade"
                            onChange={ updateState }
                            placeholder="Select Grade"
                            value={ state.grade !== null ? state.grade : '' }>
                            <option value="" disabled>Select Grade</option>
                            <option value="0">A</option>
                            <option value="1">B</option>
                            <option value="2">C</option>
                            <option value="3">D</option>
                            <option value="4">F</option>
                        </select>
                    </div>
                </div>
            </form>
            <hr className="pb-0" />
            <div className="d-flex align-items-end justify-content-end">
                <button className="btn btn-primary" disabled={ saving } onClick={ saveState }>Save</button>
            </div>
        </div>
    );
}

export default StateEdit;
