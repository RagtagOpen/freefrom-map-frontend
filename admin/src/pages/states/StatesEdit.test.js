import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import StateEdit from './StateEdit';

const mockStore = configureStore([]);

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({
        id: 1
    }),
    useLocation: () => ({
        pathname: "/states/1/edit"
    }),
    useRouteMatch: () => ({ url: '/states/1/edit' }),
}));

describe('StateEdit Component', () => {
    let store;
    let component;

    beforeEach(() => {

        store = mockStore({
            states: {
                loading: false,
                loaded: true,
                data: [
                    {
                        "id": 1,
                        "name": "State Name",
                        "abbrev": "SN",
                        "grade": 0
                    }
                ]
            }
        });

        component = renderer.create(
            <Provider store={ store }>
                <StateEdit />
            </Provider>
        );
    });

    it('should match edit view snapshot', () => {
        expect(component.toJSON()).toMatchSnapshot();
    })

    it('should match create new snapshot', () => {
        jest.mock('react-router-dom', () => ({
            ...jest.requireActual('react-router-dom'),
            useParams: () => ({}),
            useLocation: () => ({
                pathname: "/states/new"
            }),
            useRouteMatch: () => ({ url: '/states/1/edit' }),
        }));

        expect(component.toJSON()).toMatchSnapshot();
    })
})
