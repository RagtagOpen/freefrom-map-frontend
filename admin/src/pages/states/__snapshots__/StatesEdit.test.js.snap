// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateEdit Component should match create new snapshot 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
  >
    <ol
      className="breadcrumb"
    >
      <li
        className="breadcrumb-item"
      >
        <a
          href="/states"
        >
          State Scores
        </a>
      </li>
      <li
        aria-current="page"
        className="breadcrumb-item active"
      >
        State Name
      </li>
    </ol>
  </nav>
  <form>
    <div
      className="pb-4"
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="state-name-1"
        >
          Name
        </label>
        <input
          className="form-control"
          defaultValue="State Name"
          id="state-name-1"
          name="name"
          onKeyUp={[Function]}
          type="text"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="state-abbrev-1"
        >
          Abbreviation
        </label>
        <input
          className="form-control"
          defaultValue="SN"
          id="state-abbrev-1"
          name="abbrev"
          onKeyUp={[Function]}
          type="text"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="state-grade-1"
        >
          Grade
        </label>
        <select
          className="form-control"
          id="state-grade-1"
          name="grade"
          onChange={[Function]}
          placeholder="Select Grade"
          value={0}
        >
          <option
            disabled={true}
            value=""
          >
            Select Grade
          </option>
          <option
            value="0"
          >
            A
          </option>
          <option
            value="1"
          >
            B
          </option>
          <option
            value="2"
          >
            C
          </option>
          <option
            value="3"
          >
            D
          </option>
          <option
            value="4"
          >
            F
          </option>
        </select>
      </div>
    </div>
  </form>
  <hr
    className="pb-0"
  />
  <div
    className="d-flex align-items-end justify-content-end"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
    >
      Save
    </button>
  </div>
</div>
`;

exports[`StateEdit Component should match edit view snapshot 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
  >
    <ol
      className="breadcrumb"
    >
      <li
        className="breadcrumb-item"
      >
        <a
          href="/states"
        >
          State Scores
        </a>
      </li>
      <li
        aria-current="page"
        className="breadcrumb-item active"
      >
        New State
      </li>
    </ol>
  </nav>
  <form>
    <div
      className="pb-4"
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="state-name-undefined"
        >
          Name
        </label>
        <input
          className="form-control"
          id="state-name-undefined"
          name="name"
          onKeyUp={[Function]}
          type="text"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="state-abbrev-undefined"
        >
          Abbreviation
        </label>
        <input
          className="form-control"
          id="state-abbrev-undefined"
          name="abbrev"
          onKeyUp={[Function]}
          type="text"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="state-grade-undefined"
        >
          Grade
        </label>
        <select
          className="form-control"
          id="state-grade-undefined"
          name="grade"
          onChange={[Function]}
          placeholder="Select Grade"
        >
          <option
            disabled={true}
            value=""
          >
            Select Grade
          </option>
          <option
            value="0"
          >
            A
          </option>
          <option
            value="1"
          >
            B
          </option>
          <option
            value="2"
          >
            C
          </option>
          <option
            value="3"
          >
            D
          </option>
          <option
            value="4"
          >
            F
          </option>
        </select>
      </div>
    </div>
  </form>
  <hr
    className="pb-0"
  />
  <div
    className="d-flex align-items-end justify-content-end"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
    >
      Save
    </button>
  </div>
</div>
`;
